deployment/ccip/changeset/v1_6/cs_add_new_chain_e2e.go

```diff
@@ -0,0 +1,932 @@
+	out, err = ConnectNewChainChangeset.Apply(e, c.connectNewChainConfig())
+	if err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("failed to run ConnectNewChainChangeset: %w", err)
+	}
+	allProposals = append(allProposals, out.MCMSTimelockProposals...)
+
+	proposal, err := aggregateProposals(
+		e,
+		state,
+		allProposals,
+		nil,
+		fmt.Sprintf("Promote chain with selector %d for testing", c.NewChain.Selector),
+		c.MCMSConfig,
+	)
+	if err != nil {
+		return deployment.ChangesetOutput{}, fmt.Errorf("failed to build proposal: %w", err)
+	}
+
+	if proposal == nil {
+		return deployment.ChangesetOutput{}, nil
+	}
+	return deployment.ChangesetOutput{MCMSTimelockProposals: []mcmslib.TimelockProposal{*proposal}}, nil
+}
+
+///////////////////////////////////
+// END PromoteNewChainForTestingChangeset
+///////////////////////////////////
+
+///////////////////////////////////
+// START ConnectNewChainChangeset
+///////////////////////////////////
+
+// ConnectionConfig defines how a chain should connect with other chains.
+type ConnectionConfig struct {
+	// RMNVerificationDisabled is true if we do not want the RMN to bless messages FROM this chain.
+	RMNVerificationDisabled bool
+	// AllowListEnabled is true if we want an allowlist to dictate who can send messages TO this chain.
+	AllowListEnabled bool
+}
+
+// ConnectNewChainConfig is a configuration struct for ConnectNewChainChangeset.
+type ConnectNewChainConfig struct {
+	// NewChainSelector is the selector of the new chain to connect.
+	NewChainSelector uint64
+	// NewChainConnectionConfig defines how the new chain should connect with other chains.
+	NewChainConnectionConfig ConnectionConfig
+	// RemoteChains are the chains to connect the new chain to.
+	RemoteChains map[uint64]ConnectionConfig
+	// TestRouter is true if we want to connect via test routers.
+	TestRouter *bool
+	// MCMSConfig is the MCMS configuration, omit to use deployer key only.
+	MCMSConfig *changeset.MCMSConfig
+}
+
+func (c ConnectNewChainConfig) ValidateNewChain(env deployment.Environment, state changeset.CCIPOnChainState) error {
+	err := deployment.IsValidChainSelector(c.NewChainSelector)
+	if err != nil {
+		return fmt.Errorf("chain selector is invalid: %w", err)
+	}
+
+	chainState, ok := state.Chains[c.NewChainSelector]
+	if !ok {
+		return fmt.Errorf("chain with selector %d not found", c.NewChainSelector)
+	}
+
+	// When running this changeset, there is no case in which the new chain contract should be owned by MCMS,
+	// which is why we do not use MCMSConfig to determine the ownedByMCMS variable.
+	err = c.validateChain(env.GetContext(), chainState, env.Chains[c.NewChainSelector].DeployerKey.From, false)
+	if err != nil {
+		return fmt.Errorf("failed to validate chain with selector %d: %w", c.NewChainSelector, err)
+	}
+
+	return nil
+}
+
+func (c ConnectNewChainConfig) ValidateRemoteChains(env deployment.Environment, state changeset.CCIPOnChainState) error {
```

Comment:
can it be unexported?

Reply:
https://github.com/smartcontractkit/chainlink/pull/16705/commits/8104b00de8d2820f96e1d73a532a8a32c98f51f0